### PR TITLE
Expose information about configurations related to scripts

### PR DESCRIPTION
**Pull request by breskeby**
- tweaked the ResolveConfigurationDependenciesBuildOperationType.Details interface
to expose information about configuration related to projects and related to
script classpath.

As discussed a tweak in how we expose resolved configuration to allow better distinction between build script related configurations (settings, init, build.gradle, script plugin classpath) and project related configuration.

As discussed with @ldaley and @oehme in #bt-cs channel this should get into the 4.4 rc 1. 

CI: https://builds.gradle.org/project.html?projectId=Gradle_Check_Stage_BranchBuildAccept&branch_Gradle_Check_Stage_BranchBuildAccept=breskeby%2Fresolved-script-configuration-ops

### CODE REVIEW

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ResolvedConfigurationBuildOperationIntegrationTest.groovy [hunk 1]
```diff
@@ -169,4 +160,74 @@ class ResolvedConfigurationBuildOperationIntegrationTest extends AbstractHttpDep
         op.result.resolvedDependenciesCount == 1
     }
 
+    def "script classpath dependencies are exposed"() {
```
Comment (ldaley) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ResolvedConfigurationBuildOperationIntegrationTest.groovy]: We should test init scripts, settings scripts, and script plugins here.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 2]
```diff
@@ -208,8 +211,9 @@ public DefaultConfiguration(final Path identityPath, Path path, String name,
         this.attributesFactory = attributesFactory;
         this.configurationAttributes = attributesFactory.mutable();
         this.intrinsicFiles = new ConfigurationFileCollection(Specs.<Dependency>satisfyAll());
-
+        this.isScriptConfiguration = isScriptConfiguration;
```
Comment (ldaley) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: Let's extract something stronger for this.

Something like:

```
interface ConfigurationUseSite {
  String getProjectPath() // project path if owned by project
  boolean isScript()
}
```

With similar mutually exclusive semantics.

Comment (oehme) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: And make the identity path creation lazy. That's actually a nice change for performance too I believe.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ResolvedConfigurationBuildOperationIntegrationTest.groovy (hunk 1 above)
Comment (breskeby) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ResolvedConfigurationBuildOperationIntegrationTest.groovy]: added test coverage for that

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java (hunk 2 above)
Comment (breskeby) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: @ldaley fixed this. I'll make the identityPath lazy when it got this merged into master @oehme

**Comment by breskeby** [comment]
@ldaley addressed your review items

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationUseSite.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+interface ConfigurationUseSite {
+    String getProjectPath(); // project path if owned by project
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationUseSite.java]: ❌ Should be a javadoc comment

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 4]
```diff
@@ -189,7 +191,8 @@ public DefaultConfiguration(final Path identityPath, Path path, String name,
                                 Instantiator instantiator,
                                 NotationParser<Object, ConfigurablePublishArtifact> artifactNotationParser,
                                 ImmutableAttributesFactory attributesFactory,
-                                RootComponentMetadataBuilder rootComponentMetadataBuilder) {
+                                RootComponentMetadataBuilder rootComponentMetadataBuilder,
+                                ConfigurationUseSite configurationUseSite) {
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: ❌ Should replace the `identityPath`, not add yet another argument.

**Comment by breskeby** [comment]
I think `identityPath` is not reduntant here but `projectPath`. changed the ctor and made projectPath lazy

**Pull request merged**